# Churn Analysis Project 

### <u>What is churn?</u>
A customer is considered `churned` if they discontinue their telecom subscription (Churn = Yes).

### <u>Why churn matters</u>
Understanding Customer Churn is crucial for businesses to:
- Identify patterns, factors and indicators that contribute to revenue loss, as Churn directly impacts revenue stability.
- Acquiring new customers is more expensive than retaining existing ones.
- Identifying high-risk customers enables targeted retention strategies.

### The Data Dictionary

| S/N | Column Name | Description |
| :--- | :--- | :--- |
| 1 | **customerID** | A unique string identifying each customer (e.g., 7590-VHVEG). |
| 2 | **gender** | Whether the customer is a male or a female. |
| 3 | **SeniorCitizen** | A binary indicator (1 or 0) showing if the customer is aged 65 or older. |
| 4 | **Partner** | Whether the customer is married or has a domestic partner (Yes/No). |
| 5 | **Dependents** | Whether the customer lives with children, parents, or other family members they support (Yes/No). |
| 6 | **tenure** | The number of months the customer has stayed with the Telco. |
| 7 | **PhoneService** | Whether the customer is subscribed to making and receiving calls (Yes/No). |
| 8 | **MultipleLines** | Whether the customer has more than one phone line (Yes, No, or No phone service). |
| 9 | **InternetService** | The type of internet provider (DSL, Fiber optic, or No). |
| 10 | **OnlineSecurity** | Whether the customer has a firewall/malware protection service (Yes, No, or No internet service). |
| 11 | **OnlineBackup** | Whether the customer pays for cloud storage for their files (Yes, No, or No internet service). |
| 12 | **DeviceProtection** | Whether the customer has an insurance plan for devices and hardware (Yes, No, or No internet service). |
| 13 | **TechSupport** | Whether the customer has a priority technical support plan (Yes, No, or No internet service). |
| 14 | **StreamingTV** | Whether the customer uses the internet to stream television programs (Yes, No, or No internet service). |
| 15 | **StreamingMovies** | Whether the customer uses the internet to stream movies (Yes, No, or No internet service). |
| 16 | **Contract** | The contract term length between the customer and the Telco (Month-to-month, One year, Two year) |
| 17 | **PaperlessBilling** | Whether the customer gets digital invoices instead of paper ones (Yes/No). |
| 18 | **PaymentMethod** | How the customer pays for his/her various subscribed services with the Telco (Electronic check, Mailed check, Bank transfer, or Credit card). |
| 19 | **MonthlyCharges** | The amount charged to the customer every month. |
| 20 | **TotalCharges** | The total amount charged to the customer over their entire lifetime. |
| 21 | **Churn** | Whether the customer left the company within the last month (Yes/No). |

### Import needed Libraries

In [35]:
import pandas as pd

### Inspect Data

In [36]:
# Enable full column display
pd.set_option('display.max_columns', None)

churn_df = pd.read_csv(r"dataset/WA_Fn-UseC_-Telco-Customer-Churn 2.csv")

# View first 10 Rows
churn_df.head(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   
5  9305-CDSKC  Female              0      No         No       8          Yes   
6  1452-KIOVK    Male              0      No        Yes      22          Yes   
7  6713-OKOMC  Female              0      No         No      10           No   
8  7892-POOKP  Female              0     Yes         No      28          Yes   
9  6388-TABGU    Male              0      No        Yes      62          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   
5               Yes     Fiber optic             No           No   
6               Yes     Fiber optic             No          Yes   
7  No phone service             DSL            Yes           No   
8               Yes     Fiber optic             No           No   
9                No             DSL            Yes          Yes   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   
5              Yes          No         Yes             Yes  Month-to-month   
6               No          No         Yes              No  Month-to-month   
7               No          No          No              No  Month-to-month   
8              Yes         Yes         Yes             Yes  Month-to-month   
9               No          No          No              No        One year   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   
5              Yes           Electronic check           99.65        820.5   
6              Yes    Credit card (automatic)           89.10       1949.4   
7               No               Mailed check           29.75        301.9   
8              Yes           Electronic check          104.80      3046.05   
9               No  Bank transfer (automatic)           56.15      3487.95   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  
5   Yes  
6    No  
7    No  
8   Yes  
9    No

#### Check for Errors, Missen Values, Duplicates and Inconsistencies 

In [37]:
# Summary Statistics on numeric columns
churn_df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

Observations from the summary above:
- Over 75% of the customers are not senior citizens.
- Based on the percentiles, there is no strong indication of extreme outliers in tenure or MonthlyCharges.
- There may be new customers in the dataset as there are tenures of 0 Months recorded.
- The relatively high standard deviation in tenure and monthly charges indicates high variability in customer tenure and monthly charges.

In [38]:
# Summary statistics on categorical columns
churn_df.describe(include="object")

customerID gender Partner Dependents PhoneService MultipleLines  \
count         7043   7043    7043       7043         7043          7043   
unique        7043      2       2          2            2             3   
top     3186-AJIEK   Male      No         No          Yes            No   
freq             1   3555    3641       4933         6361          3390   

       InternetService OnlineSecurity OnlineBackup DeviceProtection  \
count             7043           7043         7043             7043   
unique               3              3            3                3   
top        Fiber optic             No           No               No   
freq              3096           3498         3088             3095   

       TechSupport StreamingTV StreamingMovies        Contract  \
count         7043        7043            7043            7043   
unique           3           3               3               3   
top             No          No              No  Month-to-month   
freq          3473        2810            2785            3875   

       PaperlessBilling     PaymentMethod TotalCharges Churn  
count              7043              7043         7043  7043  
unique                2                 4         6531     2  
top                 Yes  Electronic check         20.2    No  
freq               4171              2365           11  5174

Observations from the summary above:
- There are sligthly more Male than Female Customers `7043 - 3555 Males = 3488 Females`
- Most customers do not have Partners or Dependents
- About 90% of customers have phone service
- Fiber Optic is the most popular Internet Service used by the customers.
- Around 44–50% of customers do not subscribe to Online Security, Online Backup, Device Protection, or Tech Support.
- There are more customers who do not Stream TVs and Movies than those who do.
- Month-to-Month is the most popular contract type.
- About 73% of customers did not churn, while roughly 27% churned.

In [39]:
### Check for missen values
churn_df.isnull().mean()

customerID          0.0
gender              0.0
SeniorCitizen       0.0
Partner             0.0
Dependents          0.0
tenure              0.0
PhoneService        0.0
MultipleLines       0.0
InternetService     0.0
OnlineSecurity      0.0
OnlineBackup        0.0
DeviceProtection    0.0
TechSupport         0.0
StreamingTV         0.0
StreamingMovies     0.0
Contract            0.0
PaperlessBilling    0.0
PaymentMethod       0.0
MonthlyCharges      0.0
TotalCharges        0.0
Churn               0.0
dtype: float64

There are no empty cells in the dataset.

In [40]:
# Checking for duplicates
print(f"{churn_df.duplicated().sum()} Duplicates Found")

0 Duplicates Found


### Checking for Inconsistencies and Fixing of Data Types

In [41]:
# View first 5 Rows
churn_df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

In [42]:
# View info table showing Null Count and Data Type of every column in the dataset
churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


Observations:
- `TotalCharges` shouldn't be in `object` datatype, it should be in `float`
- `object` columns should be converted to `string` or `categorical` datatype

In [43]:
# Inspecting 'TotalCharges' for troublesome characters
print(
    "Found the following strings:",
    [val for val in churn_df["TotalCharges"].unique() if val.isalpha() or val.isspace()],
    sep="\n"
)

Found the following strings:
[' ']


Observations:
- No alphabetic characters were found in the `TotalCharges` column, but a **space** character was found. (This was why Pandas resorted to setting the columnn as `object` in the first place.

In [44]:
# Converting to float using pd.to_numeric and setting errors to NaN
churn_df["TotalCharges"] = pd.to_numeric(churn_df["TotalCharges"], errors="coerce")

In [45]:
# Check for null values
churn_df.isnull().sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [46]:
# Inspect the null values in the 'TotalCharges' Column
churn_df[churn_df["TotalCharges"].isnull()]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
488            No  No phone service             DSL                  Yes   
753           Yes                No              No  No internet service   
936           Yes                No             DSL                  Yes   
1082          Yes               Yes              No  No internet service   
1340           No  No phone service             DSL                  Yes   
3331          Yes                No              No  No internet service   
3826          Yes               Yes              No  No internet service   
4380          Yes                No              No  No internet service   
5218          Yes                No              No  No internet service   
6670          Yes               Yes             DSL                   No   
6754          Yes               Yes             DSL                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
488                    No                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                  Yes                   No   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                  Yes                   No                  Yes   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
488                   Yes                   No  Two year              Yes   
753   No internet service  No internet service  Two year               No   
936                   Yes                  Yes  Two year               No   
1082  No internet service  No internet service  Two year               No   
1340                  Yes                   No  Two year               No   
3331  No internet service  No internet service  Two year               No   
3826  No internet service  No internet service  Two year               No   
4380  No internet service  No internet service  Two year               No   
5218  No internet service  No internet service  One year              Yes   
6670                  Yes                   No  Two year               No   
6754                   No                   No  Two year              Yes   

                  PaymentMethod  MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)           52.55           NaN    No  
753                Mailed check           20.25           NaN    No  
936                Mailed check           80.85           NaN    No  
1082               Mailed check           25.75           NaN    No  
1340    Credit card (automatic)           56.05           NaN    No  
3331               Mailed check           19.85       

Observation:
- I assume their `TotalCharges` were kept as an empty string because of some error unknown to me, but looking closely at the table, something stands out - the `tenure` are all `0`.

A quick table to view if `tenure` of `0` have anything missen: 

In [47]:
churn_df[churn_df["tenure"] == 0]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
488            No  No phone service             DSL                  Yes   
753           Yes                No              No  No internet service   
936           Yes                No             DSL                  Yes   
1082          Yes               Yes              No  No internet service   
1340           No  No phone service             DSL                  Yes   
3331          Yes                No              No  No internet service   
3826          Yes               Yes              No  No internet service   
4380          Yes                No              No  No internet service   
5218          Yes                No              No  No internet service   
6670          Yes               Yes             DSL                   No   
6754          Yes               Yes             DSL                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
488                    No                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                  Yes                   No   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                  Yes                   No                  Yes   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
488                   Yes                   No  Two year              Yes   
753   No internet service  No internet service  Two year               No   
936                   Yes                  Yes  Two year               No   
1082  No internet service  No internet service  Two year               No   
1340                  Yes                   No  Two year               No   
3331  No internet service  No internet service  Two year               No   
3826  No internet service  No internet service  Two year               No   
4380  No internet service  No internet service  Two year               No   
5218  No internet service  No internet service  One year              Yes   
6670                  Yes                   No  Two year               No   
6754                   No                   No  Two year              Yes   

                  PaymentMethod  MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)           52.55           NaN    No  
753                Mailed check           20.25           NaN    No  
936                Mailed check           80.85           NaN    No  
1082               Mailed check           25.75           NaN    No  
1340    Credit card (automatic)           56.05           NaN    No  
3331               Mailed check           19.85       

Observation:
- The `TotalCharages` are NaN in this case because their `tenure` are `0`. To fix this, the NaN values will be replaced with their `MonthlyCharges`
This is assumed that they are new customers and their `TotalCharges` were yet to be aggregated due to poor system design.

In [48]:
# target the index of the null values
for idx in churn_df[churn_df["TotalCharges"].isnull()].index:
    # Replace the 'TotalCharges' with the 'MonthlyCharges'
    churn_df.loc[idx, "TotalCharges"] = churn_df.loc[idx, "MonthlyCharges"]

# View to confirm changes have been successfullly made
churn_df[churn_df["tenure"] == 0]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
488            No  No phone service             DSL                  Yes   
753           Yes                No              No  No internet service   
936           Yes                No             DSL                  Yes   
1082          Yes               Yes              No  No internet service   
1340           No  No phone service             DSL                  Yes   
3331          Yes                No              No  No internet service   
3826          Yes               Yes              No  No internet service   
4380          Yes                No              No  No internet service   
5218          Yes                No              No  No internet service   
6670          Yes               Yes             DSL                   No   
6754          Yes               Yes             DSL                  Yes   

             OnlineBackup     DeviceProtection          TechSupport  \
488                    No                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                  Yes                   No   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                  Yes                   No                  Yes   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
488                   Yes                   No  Two year              Yes   
753   No internet service  No internet service  Two year               No   
936                   Yes                  Yes  Two year               No   
1082  No internet service  No internet service  Two year               No   
1340                  Yes                   No  Two year               No   
3331  No internet service  No internet service  Two year               No   
3826  No internet service  No internet service  Two year               No   
4380  No internet service  No internet service  Two year               No   
5218  No internet service  No internet service  One year              Yes   
6670                  Yes                   No  Two year               No   
6754                   No                   No  Two year              Yes   

                  PaymentMethod  MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)           52.55         52.55    No  
753                Mailed check           20.25         20.25    No  
936                Mailed check           80.85         80.85    No  
1082               Mailed check           25.75         25.75    No  
1340    Credit card (automatic)           56.05         56.05    No  
3331               Mailed check           19.85       

In [49]:
# No more NULL values in the dataset
churn_df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [53]:
# Map SeniorCitizen to string
churn_df["SeniorCitizen"] = churn_df["SeniorCitizen"].map({0: "No", 1:"Yes"}).astype("string")

In [57]:
# Filter object columns
object_columns = churn_df.select_dtypes("object").columns

# Convert object columns to 'string'
churn_df[object_columns] = churn_df[object_columns].astype("string")

In [54]:
# Confirm change
churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   string 
 1   gender            7043 non-null   string 
 2   SeniorCitizen     7043 non-null   string 
 3   Partner           7043 non-null   string 
 4   Dependents        7043 non-null   string 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   string 
 7   MultipleLines     7043 non-null   string 
 8   InternetService   7043 non-null   string 
 9   OnlineSecurity    7043 non-null   string 
 10  OnlineBackup      7043 non-null   string 
 11  DeviceProtection  7043 non-null   string 
 12  TechSupport       7043 non-null   string 
 13  StreamingTV       7043 non-null   string 
 14  StreamingMovies   7043 non-null   string 
 15  Contract          7043 non-null   string 
 16  PaperlessBilling  7043 non-null   string 


In [55]:
# # View first 5 Rows
churn_df.head()

customerID  gender SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female            No     Yes         No       1           No   
1  5575-GNVDE    Male            No      No         No      34          Yes   
2  3668-QPYBK    Male            No      No         No       2          Yes   
3  7795-CFOCW    Male            No      No         No      45           No   
4  9237-HQITU  Female            No      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges  TotalCharges  \
0              Yes           Electronic check           29.85         29.85   
1               No               Mailed check           56.95       1889.50   
2              Yes               Mailed check           53.85        108.15   
3               No  Bank transfer (automatic)           42.30       1840.75   
4              Yes           Electronic check           70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

### Check Unique Values in all categorical columns
This would help reveal: 
- Spelling errors
- Missen Data
- Repeated Values
- Inconsistencies in values

In [58]:
# Loop through each categorcal value
for col in churn_df.select_dtypes("string").columns:
    # Excempt customerID
    if col != "customerID":
        print(f"-----------------------{col.upper()}")
        print(churn_df[col].unique())
        print("-----------------------\n")

-----------------------GENDER
<StringArray>
['Female', 'Male']
Length: 2, dtype: string
-----------------------

-----------------------SENIORCITIZEN
<StringArray>
['No', 'Yes']
Length: 2, dtype: string
-----------------------

-----------------------PARTNER
<StringArray>
['Yes', 'No']
Length: 2, dtype: string
-----------------------

-----------------------DEPENDENTS
<StringArray>
['No', 'Yes']
Length: 2, dtype: string
-----------------------

-----------------------PHONESERVICE
<StringArray>
['No', 'Yes']
Length: 2, dtype: string
-----------------------

-----------------------MULTIPLELINES
<StringArray>
['No phone service', 'No', 'Yes']
Length: 3, dtype: string
-----------------------

-----------------------INTERNETSERVICE
<StringArray>
['DSL', 'Fiber optic', 'No']
Length: 3, dtype: string
-----------------------

-----------------------ONLINESECURITY
<StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: string
-----------------------

-----------------------ONLINEBA

Observations:
- The inspected values are faultless
- There are no ordinal categories present

In [59]:
# Save this clean dataset for future use
churn_df.to_csv("dataset/Part1-Telco-Customer_Churn.csv", index=False)
print("CSV Saved successfully")

CSV Saved successfully
